File: Leetcode-Python/CountPrimes.py
"""
Count the number of prime numbers less than a non-negative number, n.
"""


class Solution:
    # @param {integer} n
    # @return {integer}
    def countPrimes(self, n):
        is_prime = [False, False] + [True] * (n - 2)
        i = 2
        while i * i < n:
            if is_prime[i]:
                for j in xrange(i*i, n, i):
                    is_prime[j] = False
            i += 1
        return len(filter(lambda x: x, is_prime))

File: Leetcode-Java/CountPrimes.java
import java.util.ArrayList;
import java.util.List;

/**
 * Created by yhf on 15/4/30.
 */

/*
Description:

Count the number of prime numbers less than a non-negative number, n.
*/

public class CountPrimes {
    public static int countPrimes(int n) {
        if (n <= 2) {
            return 0;
        }
        List<Integer> primes = new ArrayList<Integer>();
        primes.add(2);
        for (int i = 3; i < n; i += 2) {   // 注意，步进为2
            int sqrtI = (int) Math.sqrt(i);
            for (int j = 0; i % primes.get(j) != 0; j++) {
                if (primes.get(j) > sqrtI) {
                    primes.add(i);
                    break;
                }
            }
        }
        return primes.size();
    }

    public static void main(String[] args) {
        int i = countPrimes(499979);
    }
}
